### DISCUSSION TITLE
Domain argument for `Trusted-Set-Cookie` and `Set-Cookie` Scriptlets should be considered.

### DISCUSSION THREAD

**Discussion by ghost** [uBlockOrigin/uBlock-issues#2752] [Ideas]
@gorhill

A domain argument in the Set-Cookie scriptlets should be something to consider to add.

This is the argument for it. 
There in many/most cases, cookies will be set in a different 'domain' than what the website is setting the cookies, which means that the cookie will be duplicated unnecessarily, in different domains.
It doesn't cause any issue when setting the cookies, apparently, not I haven't encounter any, but having duplicates is not good when it comes to clearing cookies and all that.

For example, Youtube, If you use: `youtube.com##+js(trusted-set-cookie, PREF, gl=US)` the cookie will be set in `www.youtube.com`, while Youtube will set the PREF cookie in `.youtube.com`, and while 'it works', that means there will be two PREF cookies being stored in the browser, one in the `.youtube.com` and `www.youtube.com`.

That difference between domains is not so obvious now in Chromium browsers (I don't know about Firefox), I only know Brave recently added back a way to see the cookies inside the settings Chromium removed, and you would have to switch between top-level domain to see the real PREF cookie, and the 'subdomain' `www.youtube.com` to see the one being set by the scriptlet, and which is only used for the other storages.

You can see what I mean here:
<img src="https://github.com/uBlockOrigin/uBlock-issues/assets/140436204/73821cdd-0639-4d53-90a7-97fa292d97e0" width="60%">
<img src="https://github.com/uBlockOrigin/uBlock-issues/assets/140436204/eec0aa20-7267-4149-8dd3-2a7dcc31a8bc" width="60%">

So specifying the domain is a way to avoid weird behaviors and duplicates when it comes to cookies, some cookies will need, `example.com` others `.example.com` `www.example.com` `subdomain.example.com`

Anyway, after the argument why I think domain should be considered to be added by you to uBlock directly, so other users can benefit from it, I will say that I already implemented it on my side, with what I think would be the same changes you would do.
I use Brave, so what I do is modify the scriptlets myself or sideload them like `trusted-set-cookie` in their resources file to be able to use them for now. And I went ahead and re-implemented the domain argument to the scriptlet, just like I did in the past when I created my own Set-Cookie scriptlet, and then I went ahead and modified `Brave-Set-Cookie` scriptlet (deprecated in favor of yours) and now I modified both uBlock `set-cookie` to get the same functionality.

It is just adding the domain argument, so it is easy.
This is how a filter would look with all possible arguments after with my quick change:
`youtube.com##+js(trusted-set-cookie, PREF, gl=US, 1000, , .youtube.com, reload, 1)` and `youtube.com##+js(set-cookie, Test, true, , .youtube.com)`

### `set-cookie-helper.fn`:
```diff
function setCookieHelper(
    name = '',
    value = '',
    expires = '',
    path = '',
+   domain='',
    options = {},
```
```diff
if ( path === '/' ) {
        cookieParts.push('; path=/');
    }

+  if ( domain === '' ) { domain = ''; }
+  if ( domain !== '' ) {
+       cookieParts.push('; domain=', domain);
+   }
    document.cookie = cookieParts.join('');

    if ( options.reload && cookieExists(name, value) ) {
```

### `trusted-set-cookie.js`:

```diff
  (function trustedSetCookie(
    name = '',
    value = '',
    offsetExpiresSec = '',
    path = '',
+   domain = ''
) {
```
```diff
    setCookieHelper(
        name,
        value,
        expires,
        path,
+       domain,
-       getExtraArgs(Array.from(arguments), 4)
+       getExtraArgs(Array.from(arguments), 5)
    );
```

### `set-cookie.js`:

```diff
  (function setCookie(
    name = '',
    value = '',
    path = '',
+   domain = ''
) {
```

```diff
setCookieHelper(
  name,
  value,
  '',
  path,
+ domain,
- getExtraArgs(Array.from(arguments), 3)
+ getExtraArgs(Array.from(arguments), 4)
```

Hope you agree about this.
Thank you and have a good day.

**Comment by uBlock-user** [comment]
Duplicate of https://github.com/uBlockOrigin/uAssets/discussions/18598#discussioncomment-6246121

**Reply by ghost** [reply]
@uBlock-user 
I don't get how this is a duplicate.
It is not even what I am talking about, complete different worlds, the only thing that is the same is the word 'domain', but the thread you are linking doesn't have anything to do with what I am requesting here.

Read it, the thread you are linking is about setting a cookie from different domains, more specifically: FROM Google TO Youtube.
Also, which is exactly what @gorhill answered to saying it can't be done:
>adding an option to set a cookie for a **different** domain than the user is currently vising

where he answered the correct thing, that's obviously not possible, the browser will not allow to set a Github cookie from Narnia, the cookie will not be set, because it's impossible.

So how is that related to this?
My request is about is about specifying the current domain, with that I mean, for example, just changing from `www.youtube.com` to `.youtube.com`, or `brave.com` to `.brave.com` etc. 
So no, they are not 'different domains', they are the **same** domain, but displayed differently withing the browsers logic, so you can set a cookie to be used by other subdomains or not that way (as well) and that can only be done with the `domain=` in `document.cookie`.

But there is technically no difference in the way it behaves when setting to `www.` or not, only how it limits to a domain or subdomain, but the functionality is the same. What I mean is, setting a cookie like  `PREF=gl=US` in `www.youtube.com` or `.youtube.com` will not change the fact that the cookie will be set and it will work, the only thing is the cookie will be set in `www.youtube.com` and then it will get inherited by the one in `.youtube.com`. The problem here is that, by doing that, there will be a useless cookie left behind, which the browser will display like if they were in different places, instead of just correctly updating or replacing the cookie in question, because sometimes, like in youtube, the cookie is set to be used by any youtube subdomain by using the `.youtube.com`

I mean, I already implemented it and even gave the code I used for the new uBlock Orgin Scriptlets format. I am 100% sure it works, I have used it for more than a year inside Brave Adblocker and uBlock, there is no problem about it, and if setting a cookie has an invalid domain, it will not be set anyway.

You could easily copy and paste the lines I changed in Scriptlets.js and test it, and see what I mean.

So not a duplicate.... Unless we are in 1984 and 2+2=5.

**Reply by uBlock-user** [reply]
> So not a duplicate.... Unless we are in 1984 and 2+2=5.

`document.cookie` doesn't support `domain` parameter.

**Comment by ghost** [comment]
I hope @gorhill takes a look at this and understand the 'issue' in question (let's be honest, is not the biggest issue, but it can be easily added).
Would be better  to get an answer "not interested" "no" "we don't care" than being closed because of duplicate that doesn't apply here, since it is not the same scenario.

BTW, anyone can test what I mean by using `document.cookie` in Devtools as well, just to show how the browser behaves in cases like this where a browser will set cookies differently from `document.cookie`, and then how without being able to specify the domain, it can lead to orphan cookies left behind easily.

`document.cookie = "PREF=gl=US"` vs `document.cookie = "PREF=gl=US; domain=.youtube.com"`

That's basically what the domain argument will allow to do if added to uBlock scriptlets, being able to modify the **same** domain to match whatever the website wants, because if the domain is invalid, then nothing will be set anyway.

of course any website or value can be used to test, I am just using youtube because that's what I talked about in my thread because it is a 'good example', of course using `trusted-set-cookie` would make it more obvious the 'issue', but document.cookie should be enough for anyone to understand what I mean.

Adding a domain option or argument would just make things better, not worst or more complicated, just changing/adding 10 lines made a difference.

**Reply by gorhill** [reply]
[MDN](https://developer.mozilla.org/en-US/docs/web/api/document/cookie#write_a_new_cookie)

> leading dots in domain names are ignored, but browsers may decline to set the cookie containing such dots. If a domain is specified, subdomains are always included.
